dev/com.ibm.ws.transaction.context/src/com/ibm/ws/transaction/context/internal/SerialTransactionContextImpl.java

```diff
@@ -0,0 +1,158 @@
+
+import java.io.IOException;
+import java.io.NotSerializableException;
+import java.io.ObjectOutputStream;
+import java.util.concurrent.RejectedExecutionException;
+
+import javax.transaction.InvalidTransactionException;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+
+import com.ibm.tx.jta.embeddable.EmbeddableTransactionManagerFactory;
+import com.ibm.websphere.ras.annotation.Trivial;
+import com.ibm.ws.LocalTransaction.LocalTransactionCoordinator;
+import com.ibm.ws.LocalTransaction.LocalTransactionCurrent;
+import com.ibm.ws.Transaction.UOWCurrent;
+import com.ibm.ws.tx.embeddable.EmbeddableWebSphereTransactionManager;
+import com.ibm.ws.uow.embeddable.UOWManager;
+import com.ibm.ws.uow.embeddable.UOWManagerFactory;
+import com.ibm.ws.uow.embeddable.UOWToken;
+import com.ibm.wsspi.threadcontext.ThreadContext;
+
+/**
+ * A special transaction context implementation for MicroProfile Concurrency that propagates a
+ * transaction to another thread on the condition that it is only active on one thread at a time.
+ */
+public class SerialTransactionContextImpl implements ThreadContext {
+    private static final long serialVersionUID = 1;
+
+    /**
+     * Unit of work that was on the thread of execution prior to invoking the contextual task.
+     */
+    private UOWToken suspendedUOW;
+
+    private final Transaction tx;
+
+    SerialTransactionContextImpl() {
+        try {
+            tx = EmbeddableTransactionManagerFactory.getTransactionManager().getTransaction();
+        } catch (SystemException x) {
+            throw new RejectedExecutionException(x);
+        }
+    }
+
+    @Override
+    public ThreadContext clone() {
+        try {
+            SerialTransactionContextImpl copy = (SerialTransactionContextImpl) super.clone();
+            return copy;
+        } catch (CloneNotSupportedException x) {
+            throw new RuntimeException(x);
+        }
+    }
+
+    @Override
+    public void taskStarting() throws RejectedExecutionException {
+        // Suspend whatever is currently on the thread.
+        try {
+            UOWManager uowManager = UOWManagerFactory.getUOWManager();
+            suspendedUOW = uowManager.suspend();
+        } catch (com.ibm.ws.uow.embeddable.SystemException e) {
+        }
+
+        try {
+            EmbeddableTransactionManagerFactory.getTransactionManager().resume(tx);
+        } catch (InvalidTransactionException x) {
+            throw new RejectedExecutionException(x);
+        } catch (SystemException x) {
+            throw new RejectedExecutionException(x);
+        }
+    }
+
+    @Override
+    public void taskStopping() {
+        EmbeddableWebSphereTransactionManager tm = EmbeddableTransactionManagerFactory.getTransactionManager();
+        Throwable exception = null;
+
+        // Suspend the transaction that we propagated to the thread if it is still active
+        try {
+            EmbeddableTransactionManagerFactory.getTransactionManager();
```

Comment:
Is this line doing anything?

Reply:
Nope, it was accidentally left there and I'll remove it now.